File: game-server/src/main/typescript/lib/controller/rsocket/SetupController.ts

import { MapServiceClient, GameServiceServer, PlayerServiceServer, ExtrasServiceServer } from "@shared/service_rsocket_pb";
import { ExtrasService, GameService, PlayerService, MapService } from '../../service';
import { BufferEncoders } from 'rsocket-core'
import RSocketWebSocketServer from 'rsocket-websocket-server'
import { RequestHandlingRSocket } from 'rsocket-rpc-core'
import { v4 } from 'uuid';
import { ReactiveSocket, ConnectionStatus } from 'rsocket-types';
import { GameController } from './GameController';
import { PlayerController } from './PlayerController';
import { ExtrasController } from './ExtrasController';

import { Server } from 'http';

export class SetupController {
  server: Server
    constructor(
      private mapService: MapService,
      private extrasService: ExtrasService,
      private playerService: PlayerService,
      private gameService: GameService,
      app: Express.Application
    ) {
      this.server = new Server(app);
    }

    handler(socket: ReactiveSocket<any, any>) {
      const uuid = v4();
      socket.connectionStatus()
        .subscribe((cs: ConnectionStatus) => {
          if (cs.kind == 'CLOSED' || cs.kind == 'ERROR') {
            this.playerService.disconnectPlayer(uuid);
          }
        });
        try {
        const map = this.mapService.getMap();
        new MapServiceClient(socket).setup(map);
        } catch(e) {
          console.log('err', e.message);
        }
      const handler = new RequestHandlingRSocket();
    
      handler.addService('org.coinen.pacman.GameService', new GameServiceServer(new GameController(uuid, this.gameService)));
      handler.addService('org.coinen.pacman.PlayerService', new PlayerServiceServer(new PlayerController(this.playerService, uuid)));
      handler.addService('org.coinen.pacman.ExtrasService', new ExtrasServiceServer(new ExtrasController(this.extrasService)));
      return handler;
    }
    
    transport() {
      return new RSocketWebSocketServer(
        {
          server: this.server
        },
        BufferEncoders
      );
    }
}

File: game-server/src/main/java/org/coinen/reactive/pacman/controller/socket/io/SetupController.java
package org.coinen.reactive.pacman.controller.socket.io;

import java.util.function.Supplier;

import com.corundumstudio.socketio.SocketIOClient;
import io.rsocket.rpc.rsocket.RequestHandlingRSocket;
import org.coinen.reactive.pacman.service.MapService;
import org.coinen.reactive.pacman.service.PlayerService;

public class SetupController {

    final Supplier<RequestHandlingRSocket> serverRSocketSupplier;
    final MapService                       mapService;
    final PlayerService                    playerService;

    public SetupController(Supplier<RequestHandlingRSocket> requestHandlingRSocketSupplier,
        MapService service,
        PlayerService playerService) {
        serverRSocketSupplier = requestHandlingRSocketSupplier;
        mapService = service;
        this.playerService = playerService;
    }

    public void accept(SocketIOClient client) {
//        final UUID uuid = new UUID(Clock.systemUTC().millis(), ThreadLocalRandom.current().nextLong());

        client.sendEvent("setup", mapService.getMap());
    }
}
